core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala

```diff
@@ -193,45 +252,107 @@ private[spark] class CoarseMesosSchedulerBackend(
 
   override def reregistered(d: SchedulerDriver, masterInfo: MasterInfo) {}
 
+  lazy val shuffleServiceEnabled = conf.getBoolean("spark.shuffle.service.enabled", false)
+
+  lazy val shuffleServiceMem = if (shuffleServiceEnabled) {
+    conf.getInt("spark.mesos.shuffle.service.mem", 1024)
+  } else {
+    0
+  }
+
+  lazy val shuffleServiceCpu = if (shuffleServiceEnabled) {
+    conf.getInt("spark.mesos.shuffle.service.cpu", 1)
+  } else {
+    0
+  }
+
   /**
    * Method called by Mesos to offer resources on slaves. We respond by launching an executor,
    * unless we've already launched more than we wanted to.
    */
   override def resourceOffers(d: SchedulerDriver, offers: JList[Offer]) {
-    synchronized {
+    stateLock.synchronized {
       val filters = Filters.newBuilder().setRefuseSeconds(-1).build()
 
       for (offer <- offers) {
-        val slaveId = offer.getSlaveId.toString
+        val slaveId = offer.getSlaveId.getValue
         val mem = getResource(offer.getResourcesList, "mem")
         val cpus = getResource(offer.getResourcesList, "cpus").toInt
-        if (totalCoresAcquired < maxCores &&
-            mem >= MemoryUtils.calculateTotalMemory(sc) &&
-            cpus >= 1 &&
-            failuresBySlaveId.getOrElse(slaveId, 0) < MAX_SLAVE_FAILURES &&
-            !slaveIdsWithExecutors.contains(slaveId)) {
+        logTrace("Received offer id: " + offer.getId.getValue + ", cpu: " + cpus.toString +
+          ", mem: " + mem.toString)
+
+        var minCpuRequired = 1
+        var minMemRequired = MemoryUtils.calculateTotalMemory(sc)
+        var shuffleResourcesUsed = false
+
+        if (!slaveStatuses.contains(slaveId) || !slaveStatuses(slaveId).executorRunning) {
+          minCpuRequired += shuffleServiceCpu
+          minMemRequired += shuffleServiceMem
+          shuffleResourcesUsed = true
+        }
+
+        if (taskIdToSlaveId.size < executorLimit.getOrElse(Int.MaxValue) &&
+          totalCoresAcquired < maxCores &&
+          failuresBySlaveId.getOrElse(slaveId, 0) < MAX_SLAVE_FAILURES &&
+          (!slaveStatuses.contains(slaveId) || !slaveStatuses(slaveId).taskRunning) &&
+          mem >= minMemRequired &&
+          cpus >= minCpuRequired) {
```

Comment:
this if case is very hard to read. Put a few of these cases in properly named variables and chain them together?